lib/ansible/modules/windows/win_psmodule.ps1

```diff
@@ -90,97 +139,346 @@ Function Remove-Repository{
+        }
+    }
+
+    $ReturnedModuleCount = ($ReturnedModule | Measure-Object).Count
+
+    if ( $ReturnedModuleCount -eq 1 ) {
+        $ExistingModule.Exists = $true
+        $ExistingModule.Version = $ReturnedModule.FullVersion
+    }
+
+    $ExistingModule
+}
+
 Function Install-PsModule {
-    param(
-      [Parameter(Mandatory=$true)]
-      [string]$Name,
-      [string]$Repository,
-      [bool]$AllowClobber,
-      [bool]$CheckMode
+    Param(
+        [Parameter(Mandatory=$true)]
+        [String]$Name,
+        [String]$RequiredVersion,
+        [String]$MinimumVersion,
+        [String]$MaximumVersion,
+        [String]$Repository,
+        [Bool]$AllowClobber,
+        [Bool]$SkipPublisherCheck,
+        [Bool]$AllowPrerelease,
+        [Bool]$CheckMode
     )
-    if (Get-Module -Listavailable|?{$_.name -eq $Name}){
-        $result.output = "Module $($Name) already present"
+
+    $ExistingModuleBefore = Get-PsModule -Name $Name -RequiredVersion $RequiredVersion -MinimumVersion $MinimumVersion -MaximumVersion $MaximumVersion
+
+    if ( $ExistingModuleBefore.Exists ) {
+        $result.output = "Module $($Name) already present."
+        $result.module_version = $($ExistingModuleBefore.Version).ToString()
     }
-    else {      
-      try{
-        # Install NuGet Provider if needed
-        Install-NugetProvider -CheckMode $CheckMode;
+    else {
+        try{
+            # Install NuGet provider if needed
+            Install-NugetProvider -CheckMode $CheckMode
 
-        $ht = @{
-            Name      = $Name;
-            WhatIf    = $CheckMode;
-            ErrorAction = "Stop";
-            Force     = $true;
-        };
+            if ( $AllowClobber -or $AllowPrerelease ) {
+                # Update the PowerShellGet and PackageManagement modules
+                # it's required to support AllowClobber, AllowPrerelease parameters
+                Install-PackageManagement -CheckMode $CheckMode
+                Install-PowerShellGet -CheckMode $CheckMode
+            }
 
-        # If specified, use repository name to select module source
-        if ($Repository) {
-            $ht["Repository"] = "$Repository";
-        }
-
-        # Check Powershell Version (-AllowClobber was introduced in PowerShellGet 1.6.0)
-        if ("AllowClobber" -in ((Get-Command PowerShellGet\Install-Module | Select -ExpandProperty Parameters).Keys)) {
-          $ht['AllowClobber'] = $AllowClobber;
-        }
-        
-        Install-Module @ht | out-null;
-        
-        $result.output = "Module $($Name) installed"
-        $result.changed = $true
-      }
-      catch{
-        $ErrorMessage = "Problems installing $($Name) module: $($_.Exception.Message)"
-        Fail-Json $result $ErrorMessage
-      }
+            $ht = @{
+                "Name" = $Name
+                "AllowClobber" = $AllowClobber
```

Comment:
Hmmm, IMHO because
```
$allow_clobber = Get-AnsibleParam -obj $params -name "allow_clobber" -type "bool" -default $false
```
the parameter will be defined always (?).